File: 219.contains-duplicate-ii.py
#
# @lc app=leetcode id=219 lang=python3
#
# [219] Contains Duplicate II
#

# @lc code=start
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:

        last_pos_record = {}

        for idx, n in enumerate(nums):
            if n in last_pos_record:
                if abs(last_pos_record[n] - idx) <= k:
                    return True
                else:
                    last_pos_record[n] = idx
            else:
                last_pos_record[n] = idx
        
        return False

        
# @lc code=end

File: 219.contains-duplicate-ii.java
/*
 * @lc app=leetcode id=219 lang=java
 *
 * [219] Contains Duplicate II
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Map<Integer, Integer> lastPosRecords = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (lastPosRecords.containsKey(nums[i])) {
                if (Math.abs(lastPosRecords.get(nums[i]) - i) <= k) {
                    return true;
                } else {
                    lastPosRecords.put(nums[i], i);
                }
            } else {
                lastPosRecords.put(nums[i], i);
            }
        }

        return false;
    }
}
// @lc code=end

